WalletWasabi/Services/ExtPubKeyExplorer.cs

```diff
@@ -0,0 +1,70 @@
+﻿using NBitcoin;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using WalletWasabi.Backend.Models;
+
+namespace WalletWasabi.Services
+{
+	public static class ExtPubKeyExplorer
+	{
+		public static IEnumerable<BitcoinWitPubKeyAddress> GetUnusedBech32Keys(int count, bool isInternal, BitcoinExtPubKey bitcoinExtPubKey, IEnumerable<FilterModel> filters)
+		{
+			var change = isInternal ? 1u : 0u;
+			var filterArray = filters.ToArray();
+
+			const int MaxScanDepth = 1_000;
```

Comment:
My wallet has > 10,000

Reply:
No, only over 1,000.
Nevertheless we cannot assume max depth.

Reply:
> Nevertheless we cannot assume max depth.

We are not assuming a max depth here, it is just that I chose a bad name for the constant. It should be called `AddressSet` or something like that. That number is simply for limiting the size of the address set to use, and this is important because the key derivation is a slow process. 

This algorithm has a performance of O(log2 n) so, trying with 1000, 10000 or 1000000 is basically the same (they require 10, 13 and 16 iteration respectively) but generating a million addresses when the first available could be between the first few thousands is too expensive. That's why 1000 is a good number.

> My wallet has > 10,000

In that case this algorithm will not find an unused address in the first attempt [0, 1000) and we will need to call it again to search in the ranges (1000, 2000], (2000, 3000]... an so on. In pseudo code:

```c#
IEnumerable<BitcoinWitPubKeyAddress> unusedAddresses = null;
while((unusedAddresses=GetUnusedBech32Keys(blah, blah, startingFrom) == null)
    startingFrom += AddressSet;
```

Reply:
Oh, ok that sounds really good. Then concept ACK. Should I review the implementation, too?

Reply:
No yet. Now with your Concept ACK I will add the missing parts and make something more mergeable.
> Btw, I didn't test it much.

And test it, of course.